File: backend/app/services/heap.py
"""
Heap utilities
"""
from typing import TypeVar, Generic, List, Callable, Optional

T = TypeVar('T')


class MinHeap(Generic[T]):
    """Min heap implementation"""
    def __init__(self, compare: Callable[[T, T], int]):
        self._items: List[T] = []
        self._compare = compare

    def insert(self, value: T) -> None:
        """Insert a value"""
        self._items.append(value)
        self._bubble_up(len(self._items) - 1)

    def extract_min(self) -> Optional[T]:
        """Extract minimum value"""
        if not self._items:
            return None
        if len(self._items) == 1:
            return self._items.pop()

        min_val = self._items[0]
        self._items[0] = self._items.pop()
        self._bubble_down(0)
        return min_val

    def peek(self) -> Optional[T]:
        """Peek at minimum value"""
        return self._items[0] if self._items else None

    @property
    def size(self) -> int:
        """Get heap size"""
        return len(self._items)

    @property
    def is_empty(self) -> bool:
        """Check if heap is empty"""
        return len(self._items) == 0

    def _bubble_up(self, index: int) -> None:
        """Bubble up element"""
        while index > 0:
            parent = (index - 1) // 2
            if self._compare(self._items[index], self._items[parent]) >= 0:
                break
            self._items[index], self._items[parent] = self._items[parent], self._items[index]
            index = parent

    def _bubble_down(self, index: int) -> None:
        """Bubble down element"""
        while True:
            smallest = index
            left = 2 * index + 1
            right = 2 * index + 2

            if (left < len(self._items) and
                self._compare(self._items[left], self._items[smallest]) < 0):
                smallest = left

            if (right < len(self._items) and
                self._compare(self._items[right], self._items[smallest]) < 0):
                smallest = right

            if smallest == index:
                break

            self._items[index], self._items[smallest] = self._items[smallest], self._items[index]
            index = smallest


class MaxHeap(Generic[T]):
    """Max heap implementation"""
    def __init__(self, compare: Callable[[T, T], int]):
        self._items: List[T] = []
        self._compare = compare

    def insert(self, value: T) -> None:
        """Insert a value"""
        self._items.append(value)
        self._bubble_up(len(self._items) - 1)

    def extract_max(self) -> Optional[T]:
        """Extract maximum value"""
        if not self._items:
            return None
        if len(self._items) == 1:
            return self._items.pop()

        max_val = self._items[0]
        self._items[0] = self._items.pop()
        self._bubble_down(0)
        return max_val

    def peek(self) -> Optional[T]:
        """Peek at maximum value"""
        return self._items[0] if self._items else None

    @property
    def size(self) -> int:
        """Get heap size"""
        return len(self._items)

    @property
    def is_empty(self) -> bool:
        """Check if heap is empty"""
        return len(self._items) == 0

    def _bubble_up(self, index: int) -> None:
        """Bubble up element"""
        while index > 0:
            parent = (index - 1) // 2
            if self._compare(self._items[index], self._items[parent]) <= 0:
                break
            self._items[index], self._items[parent] = self._items[parent], self._items[index]
            index = parent

    def _bubble_down(self, index: int) -> None:
        """Bubble down element"""
        while True:
            largest = index
            left = 2 * index + 1
            right = 2 * index + 2

            if (left < len(self._items) and
                self._compare(self._items[left], self._items[largest]) > 0):
                largest = left

            if (right < len(self._items) and
                self._compare(self._items[right], self._items[largest]) > 0):
                largest = right

            if largest == index:
                break

            self._items[index], self._items[largest] = self._items[largest], self._items[index]
            index = largest

File: src/lib/heap.ts
/**
 * Heap utilities
 */

export class MinHeap<T> {
  private items: T[] = []
  private compare: (a: T, b: T) => number

  constructor(compare: (a: T, b: T) => number) {
    this.compare = compare
  }

  insert(value: T): void {
    this.items.push(value)
    this.bubbleUp(this.items.length - 1)
  }

  extractMin(): T | undefined {
    if (this.items.length === 0) return undefined
    if (this.items.length === 1) return this.items.pop()

    const min = this.items[0]
    this.items[0] = this.items.pop()!
    this.bubbleDown(0)
    return min
  }

  peek(): T | undefined {
    return this.items[0]
  }

  get size(): number {
    return this.items.length
  }

  get isEmpty(): boolean {
    return this.items.length === 0
  }

  private bubbleUp(index: number): void {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2)
      if (this.compare(this.items[index], this.items[parentIndex]) >= 0) break
      this.swap(index, parentIndex)
      index = parentIndex
    }
  }

  private bubbleDown(index: number): void {
    while (true) {
      let smallest = index
      const leftChild = 2 * index + 1
      const rightChild = 2 * index + 2

      if (
        leftChild < this.items.length &&
        this.compare(this.items[leftChild], this.items[smallest]) < 0
      ) {
        smallest = leftChild
      }

      if (
        rightChild < this.items.length &&
        this.compare(this.items[rightChild], this.items[smallest]) < 0
      ) {
        smallest = rightChild
      }

      if (smallest === index) break
      this.swap(index, smallest)
      index = smallest
    }
  }

  private swap(i: number, j: number): void {
    const temp = this.items[i]
    this.items[i] = this.items[j]
    this.items[j] = temp
  }
}

export class MaxHeap<T> {
  private items: T[] = []
  private compare: (a: T, b: T) => number

  constructor(compare: (a: T, b: T) => number) {
    this.compare = compare
  }

  insert(value: T): void {
    this.items.push(value)
    this.bubbleUp(this.items.length - 1)
  }

  extractMax(): T | undefined {
    if (this.items.length === 0) return undefined
    if (this.items.length === 1) return this.items.pop()

    const max = this.items[0]
    this.items[0] = this.items.pop()!
    this.bubbleDown(0)
    return max
  }

  peek(): T | undefined {
    return this.items[0]
  }

  get size(): number {
    return this.items.length
  }

  get isEmpty(): boolean {
    return this.items.length === 0
  }

  private bubbleUp(index: number): void {
    while (index > 0) {
      const parentIndex = Math.floor((index - 1) / 2)
      if (this.compare(this.items[index], this.items[parentIndex]) <= 0) break
      this.swap(index, parentIndex)
      index = parentIndex
    }
  }

  private bubbleDown(index: number): void {
    while (true) {
      let largest = index
      const leftChild = 2 * index + 1
      const rightChild = 2 * index + 2

      if (
        leftChild < this.items.length &&
        this.compare(this.items[leftChild], this.items[largest]) > 0
      ) {
        largest = leftChild
      }

      if (
        rightChild < this.items.length &&
        this.compare(this.items[rightChild], this.items[largest]) > 0
      ) {
        largest = rightChild
      }

      if (largest === index) break
      this.swap(index, largest)
      index = largest
    }
  }

  private swap(i: number, j: number): void {
    const temp = this.items[i]
    this.items[i] = this.items[j]
    this.items[j] = temp
  }
}
